### DISCUSSION TITLE
Contour winding order and source code question

### DISCUSSION THREAD

**Discussion by kkevinchou** [recastnavigation/recastnavigation#722] [Q&A] [answered]
Some  background:
I'm doing a hobby implementation of Recast in another language and would like some help understanding a bit of logic around how we construct contours - specifically with regards to the winding order and the coordinate system. Since recast creates triangles in clock wise ordering [detailed here](https://github.com/recastnavigation/recastnavigation/discussions/697) I'm under the assumption that the contour outlines are also created in clock wise order which is backed by the comment [here](https://github.com/recastnavigation/recastnavigation/blob/main/Recast/Source/RecastContour.cpp#L155).

However, if we take a look at how `dir` is updated incremented in the previous link, we can see it is incremented by 1. and if you take a look at the definition of `rcGetDirOffsetX` and `rcGetDirOffsetY` [here](https://github.com/recastnavigation/recastnavigation/blob/main/Recast/Include/Recast.h#L1253) (and used [here](https://github.com/recastnavigation/recastnavigation/blob/main/Recast/Source/RecastContour.cpp#L138)) it suggests that incrementing the direction by 1 actually causes a counter clock wise rotation (assuming a right handed coordinate system). For example, if `dir` is 0, we start off pointing along the negative X axis, and if we increment direction by `1`, we end up pointing along the positive Z axis - a counter clockwise rotation and not the clockwise one I'd expect.

Could someone clarify what I'm misunderstanding?

**Comment by kkevinchou** [comment, marked as answer]
buillt the demo, looks like the contour is built in CCW order

**Reply by grahamboree** [reply]
I wouldn't be terribly surprised if the comments around this are maybe wrong or misleading.  Sounds like you've gotten to the bottom of this, but if you want to toss some updated comments in a PR I think we'd all benefit from your investigations. :)
